### ISSUE TITLE
conda environment management

### ISSUE THREAD

**Issue by matterhorn103** [astral-sh/uv#1703]
I'm very excited by uv and the possibilities it hints at for the future of one-stop packaging and environment management.

As a scientist, though, conda is pretty much essential to many of my projects. A package manager limited to PyPI may replace pip and venv for pure Python projects for me but will always have to sit alongside conda with conda-forge.

I get that it's still very early days, but are there any ambitions/plans to truly turn Python packaging into a single effective ecosystem and unite the science and programming camps? Might uv in future also manage conda environments or use packages from conda-forge/other repos?

Or should I reign in my excitement a little?

**Comment by zanieb** [comment]
This might be out of scope for us. I'm not sure about the long-term future, but definitely in the short term. [Prefix.dev](https://prefix.dev/) / pixi is more focused on conda integration.

**Comment by matterhorn103** [comment]
Ok, thanks. I understand, but still a pity. A single Python management tool for both camps would be nice one day.

Hopefully the collaboration between you and the pixi team continues and they become complementary tools following the same standards :)

**Comment by forrestfwilliams** [comment]
Wanted to echo my support for this question as well. Many AI/ML and scientific workflows rely on Python packages that have complex C/C++ dependencies. [GDAL](https://pypi.org/project/GDAL/) is a good example of a package that is essential for many Python applications, but installing it via Pip requires that you have already installed the GDAL C library your machine.

Conda-based tools generally do a good job of managing both the Python and C/C++ dependencies, which is why many scientists use this ecosystem. Unfortunately, using a package manager that cannot manage these lower-level dependencies is a non-starter for many of us.

Would love to see the `uv` team try to come up with their own solution for this problem sometime in the future!

**Comment by sam-goodwin** [comment]
Just ran into this with a requirement to use `mambda` to use a specific version of `tensorflow` on a Mac. It would be great if UV could support conda environments.

**Comment by den-run-ai** [comment]
just had an issue that uv does not work in a conda environment, but installation is ok. uv binary is missing.

**Comment by marcelroed** [comment]
With the latest stabilizations `uv` has covered a majority of my needs for a Python packaging solution. However, the needs of Python users in the scientific space go a little further than the stated goal of being to Python what Cargo is to Rust. In scientific Python, environments are expected to include system libraries and other important packages that don't mesh well with PyPI.

Currently my environments require something like `micromamba` or `pixi` for system packages and binaries, and then `uv` to install PyPI packages that either can't be sourced from Conda repos, or are easier to manage in PyPI. For example, I might need to install a modern C++ compiler and some system libraries with conda before using `uv` to install the rest of my requirements. A key issue with this is that solving is not possible across this boundary, leading to environments that deteriorate over time, and the requirements for an install script that orders installs instead of just a single `requirements.txt` (or `environment.yml`) that can be solved.

In short, supplanting `pip` has been a brilliant beginning for `uv` to take off, but I can't see a world where scientific Python has its needs met by a solution that can only solve using packages on PyPI. Python is not Rust, and users expect an all-encompassing Python package manager to provide more than what `pip` currently does.

@zanieb, have there been any developments in thinking about expanding capabilities in this direction, or is this still out of scope? I'm partially served by `pixi` for my use-cases, but I still find just using `uv` from the terminal to be superior when dealing with the PyPI side of package management.

Thanks for the great work in this space! `uv` has been a huge boost to my morale and productivity.

**Comment by zanieb** [comment]
Thanks for the well written comment.

We care about scientific Python (several of us worked in that domain before Astral), but right now we're focused on polishing the project management experience and working on correctness. Hopefully we'll be able to figure out a good story for Conda integration, but it's not in the immediate scope.

> Thanks for the great work in this space! uv has been a huge boost to my morale and productivity.

That's great to hear <3

**Comment by shuowpro** [comment]
Is it possible to export a Conda environment YAML file? It already supports exporting a `requirements.txt` file, so adding support for exporting a Conda environment file shouldn't be difficult. As a temporary workaround, we can still export the environment file while using pip as the package manager.

**Comment by ArthurHowardMorris** [comment]
@shuowpro: Yes. With the Conda env you are interested in active run:
```
Conda env export --file environment.yml 
```

relevant to this conversation, you'll notice that anything you are using pip to install will also be listed at the bottom of this file. I think that it is this integration with `pip` is part of what @marcelroed is wishing for above.

**Comment by matterhorn103** [comment]
@zanieb related to your response in [#6989](https://github.com/astral-sh/uv/issues/6989#issuecomment-2328978747):

> I think it's okay to support if it works, but I'm not sure how well it will translate since Conda has, for example, different names for some packages. We can't actually perform a resolution with Conda.

It has felt to me for a while that the unknown correspondence between PyPI and Conda packages is the biggest thing keeping the ecosystems separate. When looking into the topic many months ago I came across Grayskull, [this thread discussing the exact problem](https://github.com/conda/grayskull/issues/168), and [this bot-generated mapping effort](https://github.com/regro/cf-graph-countyfair/blob/master/mappings/pypi/grayskull_pypi_mapping.yaml). By no means exhaustive, but I imagine some mapping like that would be a necessary starting point for Conda/PyPI interop, so figured I'd leave the links here.

**Comment by rosmur** [comment]
@matterhorn103 alternatively, is there any (technical) reason to continue using anything in the conda ecosystem? I've made the decision to never install anaconda or anything related on my new MacBook and im all in on uv - its the future.

**Comment by tpanum** [comment]
I try to avoid `conda` in favor of `uv` whenever possible. However, certain packages have so complex external dependencies its is not always feasible.

**Comment by zanieb** [comment]
@bart-maykin please open a new issue, this does not seem like the right place for that problem.

**Comment by tobiasdiez** [comment]
The following is a concrete proposal how initial conda support may look like in `uv`:
- Add support for [PEP 725 External dependencies](https://peps.python.org/pep-0725/). This allows to specify non-python dependencies in `pyproject.toml`. For example, scipy may declare
  ```toml
  [external]
  build-requires = [
    "virtual:compiler/c",
    "virtual:compiler/cpp",
    "virtual:compiler/fortran",
    "pkg:generic/ninja",
    "pkg:generic/pkg-config",
  ]
  host-requires = [
    "virtual:interface/blas",
    "virtual:interface/lapack",
  ]
  ```
  "Support" here would mean at least that one could add external dependencies via `uv add` and that `uv` can read these information from `pyproject.toml`. This should actually be straight-forward and not too involved.

  At this point, the `pyproject.toml` file contains all the information to create a conda environment, including python and non-python dependencies.

- Render these information as a conda environment file. For this, the python dependencies and the external dependencies need to be translated into the correct conda packages. As @matterhorn103 mentioned above, grayskull maintains such a mapping that could be reused for this purpose. Then the user can use the generated conda environments files via mamba/conda.

- Bonus: Streamline the process into one command. For example, `uv sync` in an activated conda environment should also update the non-python dependencies.

**Comment by xinyu-dev** [comment]
I also support this. There are some biology packages that are only conda installable.

**Comment by hoaihuongbk** [comment]
+1000 for this support. This simplifies the operational of managing the python dependencies on a large scale.

**Comment by fepegar** [comment]
> Is it possible to export a Conda environment YAML file? It already supports exporting a `requirements.txt` file, so adding support for exporting a Conda environment file shouldn't be difficult. As a temporary workaround, we can still export the environment file while using pip as the package manager.

@shuowpro, I've written [`uv2conda`](https://pypi.org/project/uv2conda/) for that. You might find it useful.

**Comment by tupui** [comment]
Adding my upvote here. We are adding some recommendations to uv in SciPy's doc but we also need to still recommend things like pixi until we have a solution for Conda.

@charliermarsh I believe support for non Python deps is around the corner?

**Comment by zanieb** [comment]
@tupui Happy to review those SciPy changes, if you want to ping me.

I don't think support for non-Python dependencies is around the corner though :)

**Comment by binbjz** [comment]
I've been using pyenv to manage python versions and miniconda and anaconda versions. I really need support for anaconda and miniconda, this is so critical.

**Comment by notatallshaw** [comment]
I think Astral are aware people would really like this issue, I don't think there's need for any more "me too" comments, add a thumbs up to the original post.

Also, for those who really need conda packages and/or environments, and are not already aware, pixi provides project management, like uv, and does both conda and standard Python packages: https://github.com/prefix-dev/pixi

**Comment by matterhorn103** [comment]
Thought I'd just drop a link here to raise awareness of a [PR](https://github.com/conda/conda/pull/14067) I have open over at the conda repo which would add support for creating a conda environment from a specification in a `pyproject.toml` file.

I like to think that it would be a good first step towards better PyPI-conda interop and would help a little in enabling projects like `uv` to achieve this feature in future.

**Comment by tupui** [comment]
> Thought I'd just drop a link here to raise awareness of a [PR](https://github.com/conda/conda/pull/14067) I have open over at the conda repo which would add support for creating a conda environment from a specification in a `pyproject.toml` file.
> 
> I like to think that it would be a good first step towards better PyPI-conda interop and would help a little in enabling projects like `uv` to achieve this feature in future.

Yeah that might be the most sensible way forward tbh. I hate to have a specific file for the conda env while the whole ecosystem of tools transitioned from having their custom file to using the pyproject. Everything conda could very well be defined there.

**Comment by MilkClouds** [comment]
I'm the whom wants to manage virtual environment in cross-project manner, and `conda` is the best solution fits to this. And I want to use this feature along with `uv`'s powerful other features.
I would be very happy if `uv` supports feature to manage virtual environment in cross-project manner.

Installing/uninstalling, or managing `conda` package is not a instant requirement for me.

**Comment by MilkClouds** [comment]
Even it's not a seems-good workaround, `pixi`'s virtual environment even can be registered as `conda` environment.
https://stackoverflow.com/questions/70851048/does-it-make-sense-to-use-conda-poetry

**Comment by chrisrodrigue** [comment]
Side note… `conda init` modifies the user PowerShell/pwsh profiles to activate the base environment on every new shell, which is *extremely* slow. uv is already on the PATH and knows where the interpreter is, so it can fix this completely.

**Comment by trungleduc** [comment]
why do people keep using `conda` while `mamba` or `pixi` can do the same job but is way faster?

**Comment by notatallshaw** [comment]
> why do people keep using `conda` while `mamba` or `pixi` can do the same job but is way faster?

This is off topic, but conda now defaults to the C++ mamba resolver engine. So the speed differences aren't as nearly big as they used to be.

That, and some users rely on the Anaconda distribution, for, convince, licensing (uv obviously can't solve that), or other enterprise service guarantees.

**Comment by charliermarsh** [comment]
> why do people keep using `conda` while `mamba` or `pixi` can do the same job but is way faster?

Let's try to keep the conversation on-topic -- this doesn't seem relevant to uv.

**Comment by trungleduc** [comment]
> So I understand the wish, but what exactly?

Lots of packages are available on `conda-forge` but not `pypi`

**Comment by matterhorn103** [comment]
> So I understand the wish, but what exactly? `pyproject.toml` is already more complete than `environment.yml`, so supporting `environment.yml` is not a good idea.

It's about several things, but speaking primarily from a sciences perspective I can off the top of my head come up with:

1. As mentioned by @trungleduc, the repositories simply contain different packages and particularly scientific ones often necessitate conda. This was historically about big ones like `numpy` and `pytorch`, these days it's more about small ones with a purely scientific audience – tools, programs, and libraries created and released as part of academic research, for example. In computational chemistry nothing ever seems to get put on PyPI.
2. Many research projects and groups of researchers use conda, many researchers have a strong preference for/investment in conda and are not going to be persuaded to switch, I imagine many places have an institute-wide policy or need for conda. At the moment collaborating in such environments is impossible with `uv`.
3. Backwards compatibility. Scientific Python projects are often not updated long-term, especially if they were produced in an academic environment, or are research projects full of scripts. Many have their environment specified exclusively in an `environment.yml`. So to use or replicate that work you are obliged to use a tool that understands conda, and that need is unlikely to go away.
4. Tutorials or package READMEs aimed at scientists will, in my experience, exclusively recommend the use of conda and installation from `conda-forge`, which means the above are not likely to change any time soon.

**Comment by fperez** [comment]
And to expand on #1 above by @matterhorn103 and @trungleduc - the key point is that many key scientific dependencies will _never_ be on PyPI b/c they have zero Python in them, yet they are used as computational engines wrapped by a Python layer. VTK or ITK are classic examples of such packages, that were a royal, spectacular pain to install prior to conda-forge offering them, but there's a lot more like it.

All of the other points @matterhorn103 makes above are also critical, and #3 is a big one. We've made big strides in convincing the scientific community to share their code and make it reproducible, but for now that's anchored in offering reasonable `environment.yml` files. For most scientists that's already a big ask and _it works_.  Supporting that community and thousands of legacy uses (often tied to published research) is extremely important, IMHO.

**Comment by fperez** [comment]
Oops, basically repeated what @mfisher87 said 😆

**Comment by Yc-Chen** [comment]
Thanks to @fperez @mfisher87 @trungleduc for the explanation!

Just add one point for uv - and also one of the reasons I did the switch - is because the latest version of pytorch [stopped official conda support](https://pytorch.org/get-started/locally/#start-locally) (and also image below). And I was very happy to see uv has already [documented](https://docs.astral.sh/uv/guides/integration/pytorch/) how to install pytorch. I wonder how this will change point (2) and (4) in the future.

<img width="809" alt="Image" src="https://github.com/user-attachments/assets/eeca49d0-a39b-4a4e-9a13-d5377ae14418" />


For point (1) that many packages do not exist on pypi, that's a broader problem than a package manager. It is also about a place to publish and host packages, and a `.conda` format instead of whl format. Since uv works with `pyproject.toml`, I searched a bit whether conda can be more compatible with it. And I found this `conda-lock` most relevant. Maybe that could be an idea to make a step towards:

https://conda.github.io/conda-lock/

**Comment by tupui** [comment]
There is a proposal of a PEP to be able to express system and other dependencies in pyproject.toml. Until that happens, conda can still decide to have a specific section in the pyproject.toml to define these extra bits. It's yaml to toml 🤷 there is no reason to not do that now.

**Comment by mfisher87** [comment]
https://peps.python.org/pep-0725/

**Comment by Yc-Chen** [comment]
https://github.com/conda/conda/issues/12462
It has been there for >1 year....

**Comment by tupui** [comment]
> https://github.com/conda/conda/issues/12462
> It has been there for >1 year....

Yes. Still something they have to do. They are the one not in-line with the rest of the ecosystem 😉

**Comment by mfisher87** [comment]
I had a discussion about this at some point with someone very familiar with the ecosystem (I think a conda forge developer?) and they raised that `pyproject.toml` is Python-specific, and `conda` is not. Supporting `pyproject.toml` would mean they would have to treat it as an optional format and continue support for the non-python-specific format.

Pixi has decided that for them, supporting two formats is worth it. I'm glad they did :) But `pixi` is at a different level of abstraction (project management) than conda (environment management), so that likely made this decision easier. I don't know if `conda` developers necessarily see this as something they "have to do", and I would agree with that. Right now, `pixi` is leading the charge on bringing excellent ergonomics to this space, and I'm so excited about that :)

**Comment by DOSull** [comment]
Definitely would like to see this - both support for the not-pure-python packages of conda (which sounds hard), but more straightforwardly (I would think) the option to put virtual environments in some central location, like `/opt/uv/lib/...` or `~/.uv/envs` or somesuch with an associated ability to run `uv activate <environment-name>` from anywhere.

Sometimes I want to quickly try out an idea and I know that the package I need exists in some environment. In `conda` I just drop into the command line and no matter where I am in the file system `conda activate <environment-name>` let's me fire up python and explore. With `uv` as currently configured I have to also remember where that environment is on the file system. That's not a huge burden, but an option to have a centralised location for environments doesn't seem like a big ask.

Similarly a centralised location gives an IDE like vscode or whatever the ability to offer all the virtual environments installed on a machine as options if your exploring some ideas in a notebook say, not working on a particular project.

**Comment by zanieb** [comment]
> the option to put virtual environments in some central location

You're looking for #1495 

> Sometimes I want to quickly try out an idea and I know that the package I need exists in some environment. In conda I just drop into the command line and no matter where I am in the file system

The non-environment centric way to do this is just `uvx --with requests,httpx python` — no need to remember the name or path of an environment.

> virtual environments take up a lot of space and scattering them all over every project folder when many projects will have common requirements seems like a bit of a waste of storage space!

To avoid confusion here: we use hard-links and copy-on-write with a shared cache for packages, each virtual environment does not consume additional space.

**Comment by DOSull** [comment]
> The non-environment centric way to do this is just (for example) uvx --with requests,httpx python — no need to remember the name or path of an environment.

This is really useful to know. I think I'd still like the option for some more permanent configurations accessible from a central location, but this is really nice for one of my use-cases.

**Comment by zanieb** [comment]
@DOSull in that case, I'd recommend just creating a project in your central location then doing `uv run --project ~/my-central-thing python`. Anyway, we'll add support for centralized environments eventually — it's just a matter of finding the time to design and implement it.

**Comment by ubaldot** [comment]
FWIW: conda is an agnostic package manager (you can e.g. install gcc, pandoc, etc.) not only for python packages. And you can install them in separate environments. Plus it really checks dependencies conflicts. Plus it aggressively check SSL stuff. I like to think it as a "light docker".

**Comment by matterhorn103** [comment]
> [@zanieb](https://github.com/zanieb) related to your response in [#6989](https://github.com/astral-sh/uv/issues/6989#issuecomment-2328978747):
> 
> > I think it's okay to support if it works, but I'm not sure how well it will translate since Conda has, for example, different names for some packages. We can't actually perform a resolution with Conda.
> 
> It has felt to me for a while that the unknown correspondence between PyPI and Conda packages is the biggest thing keeping the ecosystems separate. When looking into the topic many months ago I came across Grayskull, [this thread discussing the exact problem](https://github.com/conda/grayskull/issues/168), and [this bot-generated mapping effort](https://github.com/regro/cf-graph-countyfair/blob/master/mappings/pypi/grayskull_pypi_mapping.yaml). By no means exhaustive, but I imagine some mapping like that would be a necessary starting point for Conda/PyPI interop, so figured I'd leave the links here.

[This thread](https://github.com/conda/grayskull/issues/564) on the grayskull repo is interesting with regards to this aspect of things.

Somehow I missed at the time of that comment that the Pixi team have been maintaining their own PyPI => `conda-forge` mapping using their [`parselmouth`](https://github.com/prefix-dev/parselmouth) tool. At the moment it seems to be only in that direction, with [work on the inverse](https://github.com/prefix-dev/parselmouth/pull/33) in progress but apparently stalled. It seems that [they'd like `parselmouth` to have it](https://github.com/conda/grayskull/issues/564#issuecomment-2436849218), though, and for it to be a library/reference useable by the whole ecosystem.

**Comment by itcarroll** [comment]
For one case that @mfisher87 and @fperez describe above, where you only need conda for non-Python packages, you are currently able to make a lot of headway with separate `environment.yml` (and `conda-lock.yml`) and `pyproject.toml` (and `uv.lock`). The key to using the conda env "foo" is:

```shell
(foo) $ export UV_PROJECT_ENVIRONMENT=$CONDA_PREFIX
```

Then if "foo" does not have python
```foo
(foo) $ uv venv --allow-existing
```
works fine. If "foo" does have Python packages though, then even
```foo
(foo) $ uv sync --inexact
```
will reinstall Python packages from PyPI (even if the versions match). I wonder if it wouldn't take too much effort for uv to support keeping those existing Python packages (from conda) if they satisfy version constraints.

**Comment by jczhang02** [comment]
actually, u can use `direnv` for auto-loading the environments.

``` bash
layout_mamba_uv() {
	# Call in .envrc (direnv) with following statement:
	#    layout mamba_uv <your_environment_name>

	MAMBA_ROOT_PREFIX="${HOME}/.conda" # Make sure this points to path of your micromamba install
	MAMBA_EXE="/usr/bin/mamba"         # Should be pointing to the micromamba binary
	# evaluating the Shell in use, e.g. bash/zsh/fish
	__my_shell=$(basename ${SHELL})
	__mamba_setup="$("${MAMBA_EXE}" shell hook --shell "${__my_shell}" --root-prefix "${MAMBA_ROOT_PREFIX}" 2>/dev/null)"
	__my_env="$1"

	eval " ${__mamba_setup}"

	if [ -n "$1" ]; then
		# Explicit environment name from layout command.
		mamba activate ${__my_env}
	elif (grep -q name: environment.yml); then
		# Detect environment name from `environment.yml` file in `.envrc` directory
		mamba activate $(grep name: environment.yml | sed -e 's/name: //')
	else
		(>&2 echo No environment specified)
		exit 1
	fi

	unset __my_shell
	unset __mamba_setup
	unset __my_env

	export UV_PROJECT_ENVIRONMENT=${CONDA_PREFIX}
}

layout mamba_uv egg
export UV_PROJECT_ENVIRONMENT=${CONDA_PREFIX}
```
